File: pandascript-vs-cdp/scripts/puppeteer/retail.js
import puppeteer from "puppeteer-core";

const endpoint = process.env.BROWSER_WS ?? "ws://127.0.0.1:9222";
const browser = await puppeteer.connect(
  endpoint.startsWith("ws://")
    ? { browserWSEndpoint: endpoint }
    : { browserURL: endpoint },
);
const context = await browser.createBrowserContext();
const page = await context.newPage();

await page.goto("https://eu.gymshark.com/es-ES/collections/all-products/mens");

const products = await page.$$eval("[class*='product-card_card-wrapper']", (cards) =>
  cards.slice(0, 3).map((card) => ({
    name: card.querySelector("[class*='product-card_title'] a")?.textContent.trim(),
    url: card.querySelector("a[href*='/products/']")?.href ?? "",
  })),
);

for (const product of products) {
  await page.goto(product.url);
  await page.waitForSelector("fieldset[class*='add-to-cart_sizes']");
  product.price = parseFloat((await page.$eval(
    "[class*='product-information_price']",
    (el) => el.textContent,
  )).replace(",", "."));
  product.sizesAvailable = await page.$$eval(
    "fieldset[class*='add-to-cart_sizes'] label[class*='size_size']",
    (labels) => labels.map((l) => l.textContent.trim()),
  );
}

console.log(JSON.stringify(products));

await page.close();
await context.close();
await browser.disconnect();

File: pandascript-vs-cdp/scripts/playwright-py/retail.py
import json
import os
import re

from playwright.sync_api import sync_playwright

endpoint = os.environ.get("BROWSER_WS", "ws://127.0.0.1:9222")

with sync_playwright() as p:
    browser = p.chromium.connect_over_cdp(endpoint)
    context = browser.new_context()
    page = context.new_page()

    page.goto("https://eu.gymshark.com/es-ES/collections/all-products/mens")

    products = page.eval_on_selector_all(
        "[class*='product-card_card-wrapper']",
        """(cards) => cards.slice(0, 3).map((card) => ({
          name: card.querySelector("[class*='product-card_title'] a")?.textContent.trim(),
          url: card.querySelector("a[href*='/products/']")?.href ?? "",
        }))""",
    )

    for product in products:
        page.goto(product["url"])
        page.wait_for_selector("fieldset[class*='add-to-cart_sizes']")
        price_text = page.text_content("[class*='product-information_price']")
        product["price"] = float(re.search(r"\d+(?:\.\d+)?", price_text.replace(",", ".")).group())
        product["sizesAvailable"] = page.eval_on_selector_all(
            "fieldset[class*='add-to-cart_sizes'] label[class*='size_size']",
            "(labels) => labels.map((l) => l.textContent.trim())",
        )

    print(json.dumps(products))

    page.close()
    context.close()
    browser.close()
